cpp/src/parquet/arrow/reader_internal.h

```diff
@@ -116,11 +116,11 @@ struct ReaderContext {
   ::arrow::MemoryPool* pool;
   FileColumnIteratorFactory iterator_factory;
   bool filter_leaves;
-  std::unordered_set<int> included_leaves;
+  std::shared_ptr<std::unordered_set<int>> included_leaves;
 
   bool IncludesLeaf(int leaf_index) const {
-    return (!this->filter_leaves ||
-            (included_leaves.find(leaf_index) != included_leaves.end()));
+    const auto& to_search = *included_leaves;
```

Comment:
[CI failure](https://github.com/apache/arrow/pull/6181/checks?check_run_id=388289548#step:5:3174); even though you don't access this reference when `filter_leaves == false`, it's triggering UBSAN.
```suggestion
    if (this->filter_leaves) {
      return included_leaves->find(leaf_index) != included_leaves->end();
    }
    return true;
```